res/controllers/Reloop-Ready-scripts.js

```diff
@@ -0,0 +1,1141 @@
+    this.padModeSelectors = Array(8);
+    padModeInstances.forEach(obj => {
+        thisContainer.padModeSelectors[obj.pos] = new components.Button({
+            midi: [0x94 + index, obj.control],
+            on: ReloopReady.padColorPalette.Blue.lit,
+            off: ReloopReady.padColorPalette.Blue.dim,
+            input: makePadModeInputHandler(obj.modeInstance),
+            trigger: function() {
+                this.output(thisContainer.currentLayer === obj.modeInstance);
+            }
+        });
+    });
+    this.padModeSelectors = this.padModeSelectors.map(obj =>
+        obj === undefined ? new components.Button({}) : obj
+    );
+
+    // button can not be controlled from software.
+    // This component instance just serves as input handler for debugging.
+    this.modeButton = new components.Button({
+        midi: [index, 0x20], // shifted: [0x0C + index, 0x11], shifted led is controllable, unshifted is not.
+        input: ReloopReady.noopInputHandler
+    });
+};
+
+ReloopReady.PadContainer.prototype = Object.create(components.ComponentContainer.prototype);
+
+ReloopReady.Deck = function(index) {
+
+    const channel = index + 1;
+
+    components.Deck.call(this, channel);
+
+    const thisDeck = this;
+    const midiOn = 0x90 + index;
+
+
+    this.play = new components.PlayButton({
+        midi: [midiOn, 0x00],
+        shiftOffset: 0x10,
+        shift: function() {
+            // match behavior labelled on hardware
+            this.inKey = "reverseroll";
+            // todo fix interaction with shift button
+        },
+    });
+
+    // needs matching cuemode to be used
+    this.cue = new components.CueButton({
+        midi: [midiOn, 0x01],
+        shiftOffset: 0x04,
+    });
+
+    this.sync = new components.SyncButton({
+        midi: [midiOn, 0x02],
+        shiftOffset: 0x1,
+    }); // TODO investigate whether custom behavior is required to match controller
+
+    this.vinyl = new components.Button({
+        midi: [midiOn, 0x0F],
+        shiftOffset: -0x08,
+        type: components.Button.prototype.types.toggle,
+        shift: function() {
+            this.inKey = "slip_enabled";
+            this.outKey = this.inKey;
+            this.input = components.Button.prototype.input;
+        },
+        unshift: function() {
+            this.input = function(channel, control, value, status, _group) {
+                if (this.isPress(channel, control, value, status)) {
+                    thisDeck.jog.vinylMode = ! thisDeck.jog.vinylMode;
+                    this.output(thisDeck.jog.vinylMode);
+                }
+            };
+        },
+        trigger: function() {
+            this.output(thisDeck.jog.vinylMode);
+        }
+    });
+
+    this.jog = new components.JogWheelBasic({
+        // midiTouch: [0x90 + index, 0x06]
+        // midiWheel: [0xB0 + index, 0x06]
+        alpha: 1/8,
+        deck: channel,
+        wheelResolution: 300,
```

Comment:
I noticed that with these jogwheel settings, it is impossible to set precise cue marks / beatgrids, as already the minimal quantum of jogwheel movement will make the track move by almost 0.01 seconds (maximally zoom into the waveform to see what I mean).

This is true for both vinyl/scratch mode and non-scratch/nudge mode.

The setting of `wheelResolution: 300` and the implied `rpm: 33 + 1/3` do correctly make the spinner GUI rotate one (1) time with one (1) physical jogwheel rotation.

(Each jogwheel has 150 teeth and 150 holes, and two optical sensors, so I guess `wheelResolution: 300` is correct)

But maybe there is another setting or variable that allows for controlling the track movement with the jogwheel in scratch mode and/or non-scratch mode, while still preserving the physically correct spinner GUI movement?

Reply:
While this is valid feedback, I'm unsure how good of a solution a workaround this (essentially hardware) problem is. Though I could imagine slowing down the speed when shift is pressed. Wdyt?

Reply:
I noticed that holding SHIFT while rotating the platter allows for fine track movement. So this is review comment is resolved.